astropy/units/core.py

```diff
@@ -2132,7 +2133,8 @@ def __call__(
                     warnings.warn(msg, UnitsWarning)
                 return UnrecognizedUnit(s)
 
-        elif isinstance(s, (int, float, np.floating, np.integer)):
+        # can't use Complex from .typing because of an import loop
+        elif isinstance(s, (int, float, complex, Fraction, np.number)):
```

Comment:
Would it make sense to just check the generic `numbers.Number`? That covers `Fraction` as well as `np.number`, and `complex` is unusual enough it can be covered by it too. Indeed, one could just generically go for `isinstance(s, Number)`, since by the time we get here, we are dealing with unusual, likely not performance critical cases. Though, one could perhaps keep `float` for speed.

p.s. With that, perhaps `sanitize_scale_type` should convert input to a smaller set of numbers. Though I like duck typing and dislike `isinstance` checks...

Reply:
If it's just the large number of classes we are listing here that bothers you then we could avoid the import loop that prevents using `Complex` by moving its definition from `units.typing` to `utils.typing`. There isn't anything in its definition that is inherently restricted to `units`.

Reply:
It is more that `numbers.Number` seems to exist really for this case... Also, people can register their own number class and why should we stop them using it?

Reply:
> It is more that `numbers.Number` seems to exist really for this case... 

The big problem with `Number` is that it does not define any attributes or methods. This means that we cannot assume that multiplication of two `Number` instances is a valid operation. `numbers.Complex` would be better, but it is useless for typing purposes (my superficial understanding is that `numbers.Complex` should be a `Protocol` but cannot be implemented as such because it is a subclass of of `Number`, which would have to be a `Protocol` too, but `Numbers` can't be because it doesn't have any methods or attributes). We define our own `Complex` and `Real` so that type checkers could understand what they mean.

> Also, people can register their own number class and why should we stop them using it?

Are you talking about allowing more types before or after sanitizing? If you mean only the former then we would also need to update `sanitize_scale_type()`. If you mean the latter then we will need to add tests that the unit formatter `format_exponential_notation()` methods with the default `format_spec` can handle them. Note that we have formatters with different default `format_spec` values, which the tests would need to take into account.

Reply:
I never think with a typing hat on, but it seemed to me that if someone took the trouble to register as a `Number`, it would be their responsibility to ensure they have the right attributes and methods.

Anyway, I thought this would be a simple simplification, but clearly it really is a bit of a mess - let's just go with what you have!

Reply:
> ...it seemed to me that if someone took the trouble to register as a `Number`, it would be their responsibility to ensure they have the right attributes and methods.

The problem with `Number` is that it does not define any attributes or methods, so _everything_ trivially satisfies the definition of a `Number`.

> Anyway, I thought this would be a simple simplification, but clearly it really is a bit of a mess - let's just go with what you have!

I am not fundamentally opposed to allowing more types to initialize unit scales, but it's clear that there are gaps in our documentation and in our tests. I would prefer to fill out those gaps before we make our API more permissive because otherwise we would make the already existing gaps even wider.

Reply:
Agreed. Python's numerical tower is tricky typing-wise. Let's start with annotating & testing our current API surface.